cpg_workflows/utils.py

```diff
@@ -38,16 +38,59 @@ def read_hail(path):
     return t
 
 
-def checkpoint_hail(t, file_name: str, checkpoint_prefix: str | None = None):
-    if checkpoint_prefix:
-        path = join(checkpoint_prefix, file_name)
-        if can_reuse(path):
-            t = read_hail(str(path))
-        else:
-            t.write(str(path), overwrite=True)
-            logging.info(f'Wrote checkpoint {path}')
-            t = read_hail(str(path))
-    return t
+def checkpoint_hail(
+        t: hl.Table | hl.MatrixTable,
+        file_name: str,
+        checkpoint_prefix: str | None = None,
+        allow_reuse=False
+):
+    """
+    checkpoint method
+    provide with a path and a prefix (GCP directory, can be None)
+    allow_reuse sets whether the checkpoint can be reused - we
+    typically want to avoid reuse, as it means we're continuing a previous
+    failure from an unknown state
+
+    Args:
+        t (hl.Table | hl.MatrixTable):
+        file_name (str): name for this checkpoint
+        checkpoint_prefix (str): path to the checkpoint directory
+        allow_reuse (bool): whether to permit reuse of an existing checkpoint
+    """
+
+    # drop the schema here
+    t.describe()
+
+    if checkpoint_prefix is None:
+        return t
+
+    path = join(checkpoint_prefix, file_name)
+    if can_reuse(path) and allow_reuse:
+        logging.info(f'Re-using {path}')
+        return read_hail(path)
+
+    logging.info(f'Checkpointing {path}')
+    t = t.checkpoint(path, overwrite=True)
+
+    # throw in a repartition
+    if file_name.endswith('.mt'):
+        # estimate partitions; fall back to 1 if low row count
+        current_rows = t.count_rows()
+    else:
+        current_rows = t.count()
+
+    partitions = current_rows // 200000 or 1
+
+    # no need to alter partitions, do not shuffle
+    if t.n_partitions() == partitions:
+        return t
+
+    logging.info(
+        f'Re-partitioning {current_rows} into {partitions} partitions'
+    )
+
+    # don't name partition arg, 'partitions' for Table, 'n_partitions' for MT
+    return t.repartition(partitions, shuffle=True)
```

Comment:
If we are going to repartition, should we do that before the checkpoint is run?

Reply:
This is a holdover from the AIP implementation, where repartitioning is done during the filtering process.

In that context, each checkpoint will have fewer and fewer entries, and repartitioning is done to prevent sparse/empty partitions occurring. To estimate the required number of partitions you need to count the rows. Counting the remaining rows causes all the hail operations to trigger and then the results get thrown away and repeated, so checkpointing first and repartitioning second in that context makes sense.

Reply:
The one I'm not sure about is whether a Batch backend repartition will **always** include a checkpoint into tmp. There's some code I don't really understand there...
```
if hl.current_backend().requires_lowering:
    <do something resulting in a checkpoint in the batch scratch space>
```

but I haven't been able to find a definition for requires lowering

Reply:
ANYWAY, this data will only need to be repartitioned once, so will rethink the location of this outside of the checkpointing method so it's not hidden